core/src/core/block.rs

```diff
@@ -307,28 +307,37 @@ impl Block {
 	/// Matches any output with a potential spending input, eliminating them
 	/// from the block. Provides a simple way to compact the block. The
 	/// elimination is stable with respect to inputs and outputs order.
+    ///
+    /// NOTE: exclude coinbase from compaction process
+    /// if a block contains a new coinbase output and
+    /// is a transaction spending a previous coinbase
+    /// we do not want to compact these away
+    ///
 	pub fn compact(&self) -> Block {
-		// the chosen ones
-		let mut new_inputs = vec![];
-
-		// build a set of all output commitments
-		let mut out_set = HashSet::new();
-		for out in &self.outputs {
-			out_set.insert(out.commitment());
-		}
-		// removes from the set any hash referenced by an input, keeps the inputs that
-		// don't have a match
-		for inp in &self.inputs {
-			if !out_set.remove(&inp.commitment()) {
-				new_inputs.push(*inp);
-			}
-		}
-		// we got ourselves a keep list in that set
-		let new_outputs = self.outputs
-			.iter()
-			.filter(|out| out_set.contains(&(out.commitment())))
-			.map(|&out| out)
-			.collect::<Vec<Output>>();
+        let in_set = self.inputs
+            .iter()
+            .map(|inp| inp.commitment())
+            .collect::<HashSet<_>>();
+
+        let out_set = self.outputs
+            .iter()
+            .filter(|out| !out.features.contains(COINBASE_OUTPUT))
```

Comment:
I believe we always want to exclude coinbase outputs (and corresponding inputs) from being compacted away.
If we ever compact away a coinbase output the block will no longer validate successfully as the outputs and the kernels will no longer be consistent.
So added this additional filter to ensure we always leave coinbase inputs and outputs untouched and uncompacted.